addons/html_editor/static/src/main/toolbar/toolbar_plugin.js

```diff
@@ -365,55 +400,45 @@ export class ToolbarPlugin extends Plugin {
-            this.closeToolbar();
+    isToolbarVisible(targetedNodes, editableSelection) {
+        if (this.isMobileToolbar) {
+            return true;
         }
-    }
 
-    shouldBeVisible(selectionData, targetedNodes) {
-        const inEditable =
-            selectionData.currentSelectionIsInEditable &&
-            !selectionData.documentSelectionIsProtected &&
-            !selectionData.documentSelectionIsProtecting;
-        if (!inEditable) {
-            return false;
-        }
-        const canDisplayToolbar = this.getResource("can_display_toolbar").every((fn) =>
-            fn(this.state.namespace)
-        );
-        if (!canDisplayToolbar) {
+        const isCollapsed = editableSelection.isCollapsed;
+        if (isCollapsed) {
             return false;
         }
-        if (this.isMobileToolbar) {
+        // Only allow the toolbar to open if the selection contains visible selected characters.
+        const selectionText = editableSelection.textContent();
+        const textCleaned = selectionText.replace(/(\r\n|\n|\r|\u200B|\uFEFF)/gm, "");
+        if (textCleaned.length) {
             return true;
         }
-        const isCollapsed = selectionData.editableSelection.isCollapsed;
-        if (isCollapsed) {
-            return this.getResource("collapsed_selection_toolbar_predicate").some((fn) =>
-                fn(selectionData)
-            );
-        }
-        return !!targetedNodes.length;
-    }
-
-    shouldPreventClosing() {
-        // Should check in the document with overlays.
-        const preventClosing = document
-            .getSelection()
-            ?.anchorNode?.closest?.("[data-prevent-closing-overlay]");
-        return preventClosing?.dataset?.preventClosingOverlay === "true";
+        // Even without textContent we display the toolbar if the selection contains a <br>
+        return targetedNodes.some(
+            (node) => node.nodeType === Node.ELEMENT_NODE && node.tagName === "BR"
+        );
     }
 
-    updateNamespace(targetedNodes) {
-        const namespaces = this.getResource("toolbar_namespaces");
-        const activeNamespace = namespaces.find((ns) => ns.isApplied(targetedNodes));
-        this.state.namespace = activeNamespace?.id;
+    /**
+     * @param {SelectionData} selectionData
+     */
+    closeToolbar(selectionData = null) {
+        if (!this.overlay.isOpen) {
+            return;
+        }
+        const anchor = selectionData?.documentSelectionIsInEditable
+            ? selectionData.editableSelection?.anchorNode
+            : document.getSelection()?.anchorNode;
+        const shouldPreventClosing =
+            anchor?.closest?.("[data-prevent-closing-overlay]")?.dataset?.preventClosingOverlay ===
+            "true";
+        if (!shouldPreventClosing) {
+            this.overlay.close();
+            this.isToolbarExpanded = false;
+            this.state.namespace = null;
+        }
```

Comment:
I disagree. If I call `closeToolbar`, the toolbar should close. Either check this before calling `closeToolbar` if it's not too often* or, if it's too often, then it should be another function which would be like `updateToolbarVisibility` or something.

* could also be through a util which would be like `shouldPreventClosing(selectionData)`

Reply:
As discussed IRL, this will be done in another task. Add a comment with:
```js
// TODO: Remove data-prevent-closing.
```